Add simple leader initiator based on a lock registry

dsyer opened the pull request:
If you hold the lock, you are the leader. This simple idea gets you
a long way if there is no "native" leader initiator. E.g. you
can use this with JdbcLockRegistry.

Commit:
Add simple leader initiator based on a lock registry

If you hold the lock, you are the leader. This simple idea gets you
a long way if there is no "native" leader initiator. E.g. you
can use this with a RDBMS with JdbcLockRegistry.

Commit:
Add some docs on leader election under "endpoints"

artembilan commented:
Thank you @dsyer !

JIRA on the matter: https://jira.spring.io/browse/INT-4058

artembilan commented on spring-integration-core/src/main/java/org/springframework/integration/support/leader/LockRegistryLeaderInitiator.java:
```diff
@@ -0,0 +1,399 @@
+ * limitations under the License.
+ */
+
+package org.springframework.integration.support.leader;
```
M-m-m, do I understand correctly that you have been forced to place this class here, in this `support` sub-package, because we have a package tangle with the top-level `leader` where `SmartLifecycleRoleController` in `support` uses `leader.event`?

I don't know how to build properly dependency matrix in the IDEA, but that really feels like `SmartLifecycleRoleController` should move to the top-level `leader` package to bring `support` one closer to dependency-free of all others.

dsyer replied:
I don't really care what package it goes in. I was just trying to find something close to the existing lock support.

artembilan replied:
OK. No problem. I think we'll revise the package tangle issue when @garyrussell comes back from vacation.
Leaving that as is right now.
Thanks

artembilan commented on spring-integration-core/src/main/java/org/springframework/integration/support/leader/LockRegistryLeaderInitiator.java:
```diff
@@ -0,0 +1,399 @@
+							// should release it
+							this.lock.unlock();
+							// Give it a chance to expire.
+							Thread.sleep(LockRegistryLeaderInitiator.this.heartBeatMillis);
```
Something dangerous is going on here.
```
* If the remote lock does not expire, or if you know it interrupts the current thread
* when it expires or is broken, then you can extend the heartbeat to Long.MAX_VALUE.
*/
private long heartBeatMillis = DEFAULT_HEART_BEAT_TIME;
```
When we have `Long.MAX_VALUE` we sleep here almost forever, therefore no any leader selection more.
Actually waiting only for the interrupt action, e.g. from the `LockContext.yield()`.

That would be fine but we have `this.lock.unlock()` before the `sleep()`. So, we don't elect any more but consider ourselves as a leader, when the `lock` will be acquired by another initiator. And he is going to treat himself as a leader, too!

Am I missing anything?
Thanks

dsyer replied:
For sure if you set the heart beat to Long.MAX_VALUE you had better have a lock that interrupts the current thread when it is released or expires (I think I explained that already, but it's important, so please feel free to emphasise). That's why the default is not Long.MAX_VALUE.

The unlock before the sleep is only going to be hit if you already had the lock and then re-acquired it (so you still hold it after the unlock). It's there to keep the lock count from increasing without limit. I'm not sure I follow what you think the problem is. You only think you're the leader if you acquire the lock. If someone else manages to do that, then the heartbeat isn't working. I guess that's not impossible, so maybe that's what you mean?

artembilan replied:
Doh. I missed the fact that we re-acquire the lock and have `holdCount` there and don't unlock in the target store:
```
if (this.delegate.getHoldCount() > 1) {
	this.delegate.unlock();
	return;
}
try {
	this.mutex.delete(this.path);
}
```
And right: after using different `LockRepository` my test starts to fail as expected.

Now the algorithm becomes more robust for me :smile:

artembilan commented on spring-integration-core/src/main/java/org/springframework/integration/support/leader/LockRegistryLeaderInitiator.java:
```diff
@@ -0,0 +1,399 @@
+										LockRegistryLeaderInitiator.this, this.context,
+										LockRegistryLeaderInitiator.this.candidate.getRole());
+							}
+							this.locked = false;
```
Looks like an additional `this.lock.unlock();` won't hurt here. In case of `Exception` from the `onGranted()` action or `leaderEventPublisher`.

dsyer replied:
Yeah, I think that wouldn't hurt, and it might help.

artembilan commented:
That's all for me for now.

I hope I haven't missed anything but here is a test-case which has to fail, because we can't have two leaders from the different `JdbcLockRegistry`s (if I understand the premise of this feature correctly): https://github.com/artembilan/spring-integration/commit/306460ae02e8d8e3818859c792d4636708c1592a. Unfortunately it doesn't.

dsyer commented:
You test doesn't fail because you are using the same `LockRepository` for both candidates. If you re-write it so it works like the `JdbcLockRegistryDifferentClientTests` you will find that it fails (as expected).

Commit:
Make thread name for leader initiator unique

In case there are multiple instances in the same context we would
like to be able to spot them in the logs.

artembilan commented on spring-integration-core/src/main/java/org/springframework/integration/support/leader/LockRegistryLeaderInitiator.java:
```diff
@@ -0,0 +1,402 @@
+		@Override
+		public void yield() {
+			if (LockRegistryLeaderInitiator.this.future != null) {
+				LockRegistryLeaderInitiator.this.future.cancel(true);
```
OK. We `yield` leadership interrupting the `leaderSelector` task, but seems for me it doesn't matter that we leave the leader election.
Shouldn't we re-submit `leaderSelector` after this cancel?

dsyer replied:
My tests seems to show that simply cancelling was enough because of the infinite loop in the leader selector. (I copied the code from Patrick's hazel cast implementation in Spring Cloud, I admit.)

artembilan commented:
Please, find the polishing https://github.com/artembilan/spring-integration/commit/00da5f94f6e549a340354dedfcff54669a2df416.

Pay attention to two additions in to the `catch` when we process `onRevoked`.
And I still insist that we should have that JDBC test-case :smile: 
I fixed all your concerns there and realized that it is enough easy to re-lead after yield, so see comments for thread barriers there.

Thank you!

dsyer commented:
LGTM. I think we should add some more detailed description of failure scenarios (split brain etc.) in the docs. Reach out and ask if you want me to write something (I don't know where to put it now that the code is in 2 branches).

artembilan commented:
@dsyer , I'm merging it as is with all our agreements, of course.
Any additional changes we can and in the subsequent PRs anyway.

Thank you again!

artembilan commented:
@dsyer , this is merged into `master` as https://github.com/spring-projects/spring-integration/commit/1752204eb63c2edabde9c7fabb34cee76eb4535e.

Thank you one more time, but, please, pay attention that I was forced to rework `yield()` logic to reschedule the selector task, because after canceling an existing `FutureTask` it becomes useless and we can't `yield()` one more time.

Not sure what's going on in Hazelcast LeaderInitiator, but may be we just don't have enough test coverage there...
